addons/binding/org.openhab.binding.allplay/src/main/java/org/openhab/binding/allplay/handler/AllPlayHandler.java

```diff
@@ -0,0 +1,499 @@
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.allplay.AllPlayBindingConstants;
+import org.openhab.binding.allplay.internal.CommonSpeakerProperties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.kaizencode.tchaikovsky.AllPlay;
+import de.kaizencode.tchaikovsky.exception.AllPlayException;
+import de.kaizencode.tchaikovsky.exception.ConnectionException;
+import de.kaizencode.tchaikovsky.exception.DiscoveryException;
+import de.kaizencode.tchaikovsky.exception.SpeakerException;
+import de.kaizencode.tchaikovsky.listener.SpeakerAnnouncedListener;
+import de.kaizencode.tchaikovsky.listener.SpeakerChangedListener;
+import de.kaizencode.tchaikovsky.listener.SpeakerConnectionListener;
+import de.kaizencode.tchaikovsky.speaker.PlayState;
+import de.kaizencode.tchaikovsky.speaker.PlayState.State;
+import de.kaizencode.tchaikovsky.speaker.PlaylistItem;
+import de.kaizencode.tchaikovsky.speaker.Speaker;
+import de.kaizencode.tchaikovsky.speaker.Speaker.LoopMode;
+import de.kaizencode.tchaikovsky.speaker.Speaker.ShuffleMode;
+import de.kaizencode.tchaikovsky.speaker.VolumeRange;
+
+/**
+ * The {@link AllPlayHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Dominic Lerbs - Initial contribution
+ */
+public class AllPlayHandler extends BaseThingHandler
+        implements SpeakerChangedListener, SpeakerAnnouncedListener, SpeakerConnectionListener {
+
+    private final Logger logger = LoggerFactory.getLogger(AllPlayHandler.class);
+    private final AllPlay allPlay;
+    private final CommonSpeakerProperties speakerProperties;
+    private Speaker speaker;
+    private VolumeRange volumeRange;
+
+    private static final String ALLPLAY_THREADPOOL_NAME = "allplayHandler";
+    private ScheduledFuture<?> reconnectionJob;
+    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool(ALLPLAY_THREADPOOL_NAME);
+
+    public AllPlayHandler(Thing thing, AllPlay allPlay, CommonSpeakerProperties properties) {
+        super(thing);
+        this.allPlay = allPlay;
+        this.speakerProperties = properties;
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing AllPlay handler for speaker " + getDeviceId());
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Waiting for speaker to be discovered");
+        try {
+            allPlay.addSpeakerAnnouncedListener(this);
+            discoverSpeaker();
+        } catch (DiscoveryException e) {
+            logger.error("Unable to discover speaker", e);
+        }
+    }
+
+    /**
+     * Tries to discover the speaker which is associated with this thing.
+     */
+    public void discoverSpeaker() {
+        try {
+            logger.debug("Starting discovery for speaker " + getDeviceId());
+            allPlay.discoverSpeaker(getDeviceId());
```

Comment:
The method will start the discovery process for the speaker, but will not wait for a response and therefore should not be long running. onSpeakerAnnounced() is called by AllJoyn as soon as the speaker is discovered.
The DiscoveryException is thrown if no connection to the local AllJoyn bus exists (which is already created in the AllPlayHandlerFactory). 
I will add an updateStatus call to offline with an according message.